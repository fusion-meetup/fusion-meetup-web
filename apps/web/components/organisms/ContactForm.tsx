"use client";

import Image from "next/legacy/image";
import { useSearchParams } from "next/navigation";
import { useCallback, useEffect, useState } from "react";

import axios from "axios";
import { useForm } from "react-hook-form";
import { MdBackspace } from "react-icons/md";

import {
  ContactFormType,
  ContactFormValues,
  contactOptions,
} from "../../lib/contact";
import fusionHeart from "../../public/fusion-heart.png";
import { Alert } from "../atoms/Alert";
import { Button } from "../atoms/Button";
import { CharCount } from "../atoms/CharCount";
import { FormFieldLabel } from "../atoms/FormFieldLabel";
import { Heading } from "../atoms/Heading";
import { Input } from "../atoms/Input";
import { Select } from "../atoms/Select";

const MAX_MESSAGE_LENGTH = 1000;

export const ContactForm: React.FC = () => {
  const searchParams = useSearchParams();
  const prefillType = searchParams?.get("t") ?? null;

  const [submitSuccess, setSubmitSuccess] = useState<boolean | undefined>();
  const [sendingFormResult, setSendingFormResult] = useState<boolean>(false);
  const [showMessageTooLongError, setShowMessageTooLongError] =
    useState<boolean>(false);

  const { register, handleSubmit, setValue, reset, watch } =
    useForm<ContactFormValues>({
      defaultValues: {
        name: "",
        email: "",
        type: "general",
        message: "",
        query: "",
      },
    });

  const contactFormSubmit = useCallback(async (data: ContactFormValues) => {
    if (data.message.length > MAX_MESSAGE_LENGTH) {
      setShowMessageTooLongError(true);
      return;
    }
    setShowMessageTooLongError(false);

    try {
      setSendingFormResult(true);

      const response = await axios.post<ApiContactResponse>(
        "/api/contact",
        data,
      );
      setSubmitSuccess(response.data.success);
    } catch (e) {
      console.error(e);
      setSubmitSuccess(false);
    } finally {
      setSendingFormResult(false);
    }
  }, []);

  // Set contact type using URL query if set
  useEffect(() => {
    if (prefillType && contactOptions.find((co) => co.value === prefillType)) {
      setValue("type", prefillType as ContactFormType);
    }
  }, [prefillType, setValue]);

  return (
    <>
      <Heading level={2}>Contact the Fusion Team</Heading>

      <div className="flex flex-row flex-wrap items-center gap-1 py-4">
        <p className="w-full">
          This contact form goes directly to the Fusion team
        </p>
        <p>We aim to reply via email as soon as possible</p>
        <Image
          alt="Fusion heart"
          src={fusionHeart}
          layout="fixed"
          width={20}
          height={20}
          className="inline-block"
        />
      </div>

      {submitSuccess !== true ? (
        <form
          onSubmit={handleSubmit(contactFormSubmit)}
          className="flex flex-col gap-6 py-4"
        >
          <div>
            <FormFieldLabel htmlFor="name">Name</FormFieldLabel>
            <Input
              {...register("name", { required: true })}
              placeholder="Arthur Dent"
              disabled={sendingFormResult}
            />
          </div>

          {/* Bot field */}
          <div className="hidden">
            <FormFieldLabel htmlFor="query">Query</FormFieldLabel>
            <Input
              {...register("query")}
              placeholder="Tell us more"
              disabled={sendingFormResult}
            />
          </div>

          <div>
            <FormFieldLabel htmlFor="email">Email</FormFieldLabel>
            <Input
              {...register("email", { required: true })}
              type="email"
              placeholder="arthur@dent.io"
              disabled={sendingFormResult}
            />
          </div>

          <div>
            <FormFieldLabel htmlFor="type">Subject</FormFieldLabel>
            <Select
              {...register("type", { required: true })}
              options={contactOptions}
              disabled={sendingFormResult}
            />
          </div>

          <div className="relative">
            <FormFieldLabel htmlFor="type">Message</FormFieldLabel>
            <Input
              multiLine
              {...register("message", { required: true })}
              placeholder="Hi Fusion team!"
              disabled={sendingFormResult}
            />

            <div className="absolute right-4 bottom-4">
              <CharCount value={watch("message")} max={MAX_MESSAGE_LENGTH} />
            </div>
          </div>

          {submitSuccess === false ? (
            <Alert type="error">
              Something went wrong – please try again, or shoot us a message on{" "}
              <a
                href="https://twitter.com/thefusion_hub"
                className="underline"
                target="_blank"
                rel="noopener noreferrer"
              >
                Twitter
              </a>{" "}
              if this seems to be broken!
            </Alert>
          ) : null}

          {showMessageTooLongError ? (
            <Alert type="error">
              Your message is too long – please shorten it to{" "}
              {MAX_MESSAGE_LENGTH} characters or less
            </Alert>
          ) : null}

          <div className="flex flex-row items-center justify-end gap-8">
            {sendingFormResult ? (
              <div className="opacity-50">Sending...</div>
            ) : (
              <Button
                typeSubmit
                value="Submit"
                color="pink"
                className="px-16"
              />
            )}
          </div>
        </form>
      ) : (
        <div className="flex flex-col gap-8 py-4">
          <Alert type="success">
            Contact form sent! Thanks for getting in touch
          </Alert>

          <Button
            color="blue"
            onClick={() => {
              reset();
              setSubmitSuccess(undefined);
            }}
          >
            <MdBackspace />
            Reset
          </Button>
        </div>
      )}
    </>
  );
};

type ApiContactResponse = {
  success: boolean;
  message: string;
};
