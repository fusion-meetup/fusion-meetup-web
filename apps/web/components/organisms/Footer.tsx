import Image from "next/legacy/image";

import { cn } from "@ui/lib/utils";

import { Socials } from "../molecules/Socials";

interface FooterProps extends React.HTMLProps<HTMLDivElement> {
  noBg?: boolean;
}

export const Footer: React.FC<FooterProps> = ({
  className,
  noBg,
  ...props
}) => {
  return (
    <footer
      {...props}
      className={cn(
        "z-2 mt-auto",
        { "bg-slate-800 text-slate-300": !noBg },
        className,
      )}
    >
      <div className="container mx-auto flex flex-row flex-wrap gap-x-4 p-4">
        <div className="flex flex-row flex-wrap items-center gap-1 py-2">
          Made with
          <Image
            alt="Fusion heart"
            src="/fusion-heart.png"
            layout="fixed"
            width={20}
            height={20}
          />
          by Fusion in Birmingham
        </div>

        <div className="ml-auto flex flex-row items-center justify-end">
          <Socials />
        </div>
      </div>
    </footer>
  );
};
