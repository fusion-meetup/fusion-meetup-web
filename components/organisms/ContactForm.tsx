import { useCallback, useEffect, useState } from "react";
import Image from "next/legacy/image";
import { useRouter } from "next/router";
import { useForm } from "react-hook-form";
import axios from "axios";
import { MdBackspace } from "react-icons/md";

import {
  ContactFormType,
  ContactFormValues,
  contactOptions,
} from "../../lib/contact";
import { Heading } from "../atoms/Heading";
import { Input } from "../atoms/Input";
import { Select } from "../atoms/Select";
import { Button } from "../atoms/Button";
import { Alert } from "../atoms/Alert";
import { FormFieldLabel } from "../atoms/FormFieldLabel";
import fusionHeart from "../../public/fusion-heart.png";

export const ContactForm: React.FC = () => {
  const router = useRouter();

  const [submitSuccess, setSubmitSuccess] = useState<boolean | undefined>();
  const [sendingFormResult, setSendingFormResult] = useState<boolean>(false);

  const { register, handleSubmit, setValue, reset } =
    useForm<ContactFormValues>({
      defaultValues: {
        name: "",
        email: "",
        type: "general",
        message: "",
        query: "",
      },
    });

  const contactFormSubmit = useCallback(async (data: ContactFormValues) => {
    try {
      setSendingFormResult(true);

      const response = await axios.post<ApiContactResponse>(
        "/api/contact",
        data
      );
      setSubmitSuccess(response.data.success);
    } catch (e) {
      setSubmitSuccess(false);
    } finally {
      setSendingFormResult(false);
    }
  }, []);

  // Set contact type using URL query if set
  useEffect(() => {
    if (contactOptions.find((co) => co.value === router.query.t)) {
      setValue("type", router.query.t as ContactFormType);
    }
  }, [router.isReady, setValue, router.query.t]);

  return (
    <>
      <Heading level={2}>Contact the Fusion Team</Heading>

      <div className="flex flex-row flex-wrap items-center gap-1 py-4">
        <p className="w-full">
          This contact form goes directly to the Fusion team
        </p>
        <p>We aim to reply via email as soon as possible</p>
        <Image
          alt="Fusion heart"
          src={fusionHeart}
          layout="fixed"
          width={20}
          height={20}
          className="inline-block"
        />
      </div>

      {submitSuccess !== true ? (
        <form
          onSubmit={handleSubmit(contactFormSubmit)}
          className="flex flex-col gap-6 py-4"
        >
          <div>
            <FormFieldLabel htmlFor="name">Name</FormFieldLabel>
            <Input
              {...register("name", { required: true })}
              placeholder="Arthur Dent"
              disabled={sendingFormResult}
            />
          </div>

          {/* Bot field */}
          <div className="hidden">
            <FormFieldLabel htmlFor="query">Query</FormFieldLabel>
            <Input
              {...register("query")}
              placeholder="Tell us more"
              disabled={sendingFormResult}
            />
          </div>

          <div>
            <FormFieldLabel htmlFor="email">Email</FormFieldLabel>
            <Input
              {...register("email", { required: true })}
              type="email"
              placeholder="arthur@dent.io"
              disabled={sendingFormResult}
            />
          </div>

          <div>
            <FormFieldLabel htmlFor="type">Subject</FormFieldLabel>
            <Select
              {...register("type", { required: true })}
              options={contactOptions}
              disabled={sendingFormResult}
            />
          </div>

          <div>
            <FormFieldLabel htmlFor="type">Message</FormFieldLabel>
            <Input
              multiLine
              {...register("message", { required: true })}
              placeholder="Hi Fusion team!"
              disabled={sendingFormResult}
            />
          </div>

          {submitSuccess === false ? (
            <Alert type="error">
              Something went wrong â€“ please try again, or shoot us a message on{" "}
              <a
                href="https://twitter.com/thefusion_hub"
                className="underline"
                target="_blank"
                rel="noopener noreferrer"
              >
                Twitter
              </a>{" "}
              if this seems to be broken!
            </Alert>
          ) : null}

          <div className="flex flex-row justify-end items-center gap-8">
            {sendingFormResult ? (
              <div className="opacity-50">Sending...</div>
            ) : (
              <Button
                typeSubmit
                value="Submit"
                color="pink"
                className="px-16"
              />
            )}
          </div>
        </form>
      ) : (
        <div className="py-4 flex flex-col gap-8">
          <Alert type="success">
            Contact form sent! Thanks for getting in touch
          </Alert>

          <Button
            color="blue"
            onClick={() => {
              reset();
              setSubmitSuccess(undefined);
            }}
          >
            <MdBackspace />
            Reset
          </Button>
        </div>
      )}
    </>
  );
};

type ApiContactResponse = {
  success: boolean;
  message: string;
};
