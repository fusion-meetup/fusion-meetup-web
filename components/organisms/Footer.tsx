import Image from "next/image";

import { Socials } from "../molecules/Socials";
import fusionHeart from "../../public/fusion-heart.png";

export const Footer: React.FC = () => {
  return (
    <footer className="z-2 bg-slate-100 dark:bg-slate-800 text-slate-600 dark:text-slate-300 mt-auto">
      <div className="container mx-auto p-4 flex flex-row flex-wrap gap-x-4">
        <div className="flex flex-row flex-wrap items-center gap-1 py-2">
          Made with
          <Image
            alt="Fusion heart"
            src={fusionHeart}
            layout="fixed"
            width={20}
            height={20}
          />
          by Fusion in Birmingham
        </div>

        <div className="ml-auto flex flex-row items-center justify-end">
          <Socials />
        </div>
      </div>
    </footer>
  );
};
