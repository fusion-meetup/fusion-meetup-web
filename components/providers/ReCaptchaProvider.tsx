import { GoogleReCaptchaProvider } from "react-google-recaptcha-v3";

const recaptchaSiteKey = process.env.NEXT_PUBLIC_RECAPTCHA_SITE_KEY;

export const ReCaptchaProvider: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  if (!recaptchaSiteKey) return <>{children}</>;

  return (
    <GoogleReCaptchaProvider reCaptchaKey={recaptchaSiteKey}>
      {children}
    </GoogleReCaptchaProvider>
  );
};
