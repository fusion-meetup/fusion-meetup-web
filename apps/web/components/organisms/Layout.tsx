import Head from "next/head";

import { cn } from "@ui/lib/utils";

import { Header } from "../molecules/Header";
import { Hero } from "../molecules/Hero";
import { Footer } from "./Footer";

interface LayoutProps {
  children: React.ReactNode;
  title?: string;
  withHero?: boolean;
  fancyBackground?: "colours";
  className?: string;
}

export const Layout: React.FC<LayoutProps> = ({
  children,
  title: pageTitle,
  withHero,
  fancyBackground,
  className,
}) => {
  const description =
    "Fusion is a Birmingham-based tech meetup, bringing the community together over great food for a social evening of tech talks aimed to inspire and educate";

  return (
    <>
      <Head>
        <title>{`Fusion Meetup â€“ ${pageTitle || description}`}</title>
        <meta name="description" content={`Fusion Meetup: ${description}`} />
      </Head>

      <div
        className={cn("flex min-h-screen flex-col", {
          "background-fusion-colours": fancyBackground === "colours",
        })}
      >
        {withHero ? <Hero /> : null}

        <Header />

        <main
          className={cn(className, "z-1 relative pb-8", {
            "pt-12": !withHero,
            "-mt-8": withHero,
          })}
        >
          {children}
        </main>

        <Footer />
      </div>
    </>
  );
};
