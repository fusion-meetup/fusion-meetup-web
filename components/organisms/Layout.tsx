import { useEffect } from "react";
import { useRouter } from "next/router";
import clsx from "clsx";
import Head from "next/head";
import Image from "next/image";

import { Header } from "../molecules/Header";
import { Footer } from "./Footer";
import fusionColoursBackground from "../../public/fusion-colours-background.png";
import { Hero } from "../molecules/Hero";
import { getBlogPostsSlugs } from "../../lib/cms/queries";

interface LayoutProps {
  children: React.ReactNode;
  title?: string;
  withHero?: boolean;
  fancyBackground?: "colours";
  className?: string;
}

export const Layout: React.FC<LayoutProps> = ({
  children,
  title,
  withHero,
  fancyBackground,
  className,
}) => {
  const router = useRouter();

  useEffect(() => {
    if (!router.isReady) return;
    if (window.location.hostname !== "localhost") return;

    const surprise = async () => {
      const slugs = await getBlogPostsSlugs();
      if (!slugs.includes("surprise")) return;

      setTimeout(() => {
        router.push("https://www.youtube.com/watch?v=dQw4w9WgXcQ");
      }, 10000);
    };

    surprise();
  }, [router, router.isReady]);

  return (
    <>
      <Head>
        <title>
          {`Fusion Meetup â€“ ${
            title || "A quarterly tech meetup held in Birmingham city centre"
          }`}
        </title>
        <meta
          name="description"
          content="Fusion Meetup: A quarterly tech meetup held in Birmingham city centre"
        />
      </Head>

      <div className="min-h-screen flex flex-col">
        {fancyBackground === "colours" && (
          <div
            className={clsx("absolute top-12 left-0 w-full", {
              "sm:top-0 md:-top-16 lg:-top-24 xl:-top-32": !withHero,
            })}
          >
            <Image
              src={fusionColoursBackground}
              alt="background image"
              width={512}
              height={741}
              layout="responsive"
              className={clsx("w-full blur-lg opacity-70 md:opacity-40", {
                "opacity-25 md:opacity-25": withHero,
              })}
            />
          </div>
        )}

        {withHero ? <Hero /> : null}

        <Header />

        <main
          className={clsx(className, "relative z-1 pb-8", {
            "pt-12": !withHero,
            "-mt-8": withHero,
          })}
        >
          {children}
        </main>

        <Footer />
      </div>
    </>
  );
};
