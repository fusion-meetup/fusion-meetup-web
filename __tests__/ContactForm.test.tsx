import { render } from "@testing-library/react";
import axios from "axios";

import { ContactForm } from "../components/organisms/ContactForm";
import { ReCaptchaProvider } from "../components/providers/ReCaptchaProvider";

jest.mock("axios");
const mockedAxios = axios as jest.Mocked<typeof axios>;

jest.mock("next/router", () => ({
  useRouter: () => ({
    isReady: true,
    query: { t: "talkSubmission" },
  }),
}));

mockedAxios.get.mockImplementation(() => Promise.resolve(true));

const renderContactForm = () => {
  return render(
    <ReCaptchaProvider>
      <ContactForm />
    </ReCaptchaProvider>
  );
};

describe("ContactForm", () => {
  it("should render", async () => {
    const { findByText } = renderContactForm();

    expect(
      await findByText("This contact form goes directly to the Fusion team")
    ).toBeInTheDocument();
  });
});
