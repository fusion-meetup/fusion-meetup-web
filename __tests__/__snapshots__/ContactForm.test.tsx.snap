// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactForm should render 1`] = `
<DocumentFragment>
  <h2
    class="break-words font-bold text-3xl md:text-4xl"
  >
    Contact the Fusion Team
  </h2>
  <div
    class="flex flex-row flex-wrap items-center gap-1 py-4"
  >
    <p
      class="w-full"
    >
      This contact form goes directly to the Fusion team
    </p>
    <p>
      We aim to reply via email as soon as possible
    </p>
    <span
      style="box-sizing: border-box; display: inline-block; overflow: hidden; width: 20px; height: 20px; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
    >
      <img
        alt="Fusion heart"
        class="inline-block"
        data-nimg="fixed"
        decoding="async"
        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
        style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
      />
      <noscript />
    </span>
  </div>
  <form
    class="flex flex-col gap-6 py-4"
  >
    <div>
      <label
        class="text-sm font-medium opacity-60 block pb-1"
        for="name"
      >
        Name
      </label>
      <input
        class="w-full px-4 py-2 h-12 border border-gray-300 rounded-md focus:outline-none focus:ring focus:ring-blue focus:border-transparent bg-white dark:bg-slate-800 dark:border-slate-700 dark:text-white disabled:opacity-50 disabled:cursor-not-allowed focus:placeholder-transparent placeholder:opacity-75 h-12"
        name="name"
        placeholder="Arthur Dent"
      />
    </div>
    <div
      class="hidden"
    >
      <label
        class="text-sm font-medium opacity-60 block pb-1"
        for="query"
      >
        Query
      </label>
      <input
        class="w-full px-4 py-2 h-12 border border-gray-300 rounded-md focus:outline-none focus:ring focus:ring-blue focus:border-transparent bg-white dark:bg-slate-800 dark:border-slate-700 dark:text-white disabled:opacity-50 disabled:cursor-not-allowed focus:placeholder-transparent placeholder:opacity-75 h-12"
        name="query"
        placeholder="Tell us more"
      />
    </div>
    <div>
      <label
        class="text-sm font-medium opacity-60 block pb-1"
        for="email"
      >
        Email
      </label>
      <input
        class="w-full px-4 py-2 h-12 border border-gray-300 rounded-md focus:outline-none focus:ring focus:ring-blue focus:border-transparent bg-white dark:bg-slate-800 dark:border-slate-700 dark:text-white disabled:opacity-50 disabled:cursor-not-allowed focus:placeholder-transparent placeholder:opacity-75 h-12"
        name="email"
        placeholder="arthur@dent.io"
        type="email"
      />
    </div>
    <div>
      <label
        class="text-sm font-medium opacity-60 block pb-1"
        for="type"
      >
        Subject
      </label>
      <select
        class="w-full h-12 px-3 py-3 border border-gray-300 rounded-md focus:outline-none focus:ring focus:ring-blue focus:border-transparent bg-white dark:bg-slate-800 dark:border-slate-700 dark:text-white disabled:opacity-50 disabled:cursor-not-allowed"
        name="type"
      >
        <option
          value="general"
        >
          General
        </option>
        <option
          value="talkSubmission"
        >
          Talk Submission
        </option>
        <option
          value="question"
        >
          Question
        </option>
        <option
          value="sponsorship"
        >
          Sponsorship
        </option>
        <option
          value="feedback"
        >
          Feedback
        </option>
      </select>
    </div>
    <div
      class="relative"
    >
      <label
        class="text-sm font-medium opacity-60 block pb-1"
        for="type"
      >
        Message
      </label>
      <textarea
        class="w-full px-4 py-2 h-12 border border-gray-300 rounded-md focus:outline-none focus:ring focus:ring-blue focus:border-transparent bg-white dark:bg-slate-800 dark:border-slate-700 dark:text-white disabled:opacity-50 disabled:cursor-not-allowed focus:placeholder-transparent placeholder:opacity-75 h-40"
        name="message"
        placeholder="Hi Fusion team!"
      />
      <div
        class="absolute bottom-4 right-4"
      />
    </div>
    <div
      class="flex flex-row justify-end items-center gap-8"
    >
      <input
        class="cursor-pointer font-bold py-2 px-4 rounded-md text-md box-content margin-0 leading-normal block w-fit flex flex-row items-center justify-center gap-2 text-base bg-pink-500 hover:bg-pink-400 outline-pink-500/50 text-black px-16 transition-all ease-out duration-100 hover:shadow focus:outline active:outline outline-3 outline-offset-2 focus:scale-95"
        type="submit"
        value="Submit"
      />
    </div>
  </form>
</DocumentFragment>
`;
